examples/wakaama/lwm2mclient.c

```diff
@@ -0,0 +1,698 @@
+            case STATE_INITIAL:
+                fprintf(stdout, "STATE_INITIAL\r\n");
+                break;
+            case STATE_BOOTSTRAP_REQUIRED:
+                fprintf(stdout, "STATE_BOOTSTRAP_REQUIRED\r\n");
+                break;
+            case STATE_BOOTSTRAPPING:
+                fprintf(stdout, "STATE_BOOTSTRAPPING\r\n");
+                break;
+            case STATE_REGISTER_REQUIRED:
+                fprintf(stdout, "STATE_REGISTER_REQUIRED\r\n");
+                break;
+            case STATE_REGISTERING:
+                fprintf(stdout, "STATE_REGISTERING\r\n");
+                break;
+            case STATE_READY:
+                fprintf(stdout, "STATE_READY\r\n");
+                break;
+            default:
+                fprintf(stdout, "Unknown...\r\n");
+                break;
+        }
+        /*
+         * This function does two things:
+         *  - first it does the work needed by liblwm2m (eg. (re)sending some
+         * packets).
+         *  - Secondly it adjusts the timeout value (default 60s) depending on the
+         * state of the transaction
+         *    (eg. retransmission) and the time between the next operation
+         */
+        DEBUG("lwm2m_step()\n");
+        result = lwm2m_step(lwm2mH, &tv);
+        fprintf(stdout, " -> State: ");
+        switch (lwm2mH->state) {
+            case STATE_INITIAL:
+                fprintf(stdout, "STATE_INITIAL\r\n");
+                break;
+            case STATE_BOOTSTRAP_REQUIRED:
+                fprintf(stdout, "STATE_BOOTSTRAP_REQUIRED\r\n");
+                break;
+            case STATE_BOOTSTRAPPING:
+                fprintf(stdout, "STATE_BOOTSTRAPPING\r\n");
+                break;
+            case STATE_REGISTER_REQUIRED:
+                fprintf(stdout, "STATE_REGISTER_REQUIRED\r\n");
+                break;
+            case STATE_REGISTERING:
+                fprintf(stdout, "STATE_REGISTERING\r\n");
+                break;
+            case STATE_READY:
+                fprintf(stdout, "STATE_READY\ren");
+                break;
+            default:
+                fprintf(stdout, "Unknown...\r\n");
+                break;
+        }
+        if (result != 0) {
+            fprintf(stderr, "lwm2m_step() failed: 0x%X\r\n", result);
+#ifdef LWM2M_BOOTSTRAP
+            if (previousState == STATE_BOOTSTRAPPING) {
+#ifdef WITH_LOGS
+                fprintf(stdout,
+                        "[BOOTSTRAP] restore security and server objects\r\n");
+#endif          /* WITH_LOGS */
+                prv_restore_objects(lwm2mH);
+                lwm2mH->state = STATE_INITIAL;
+            }
+            else
+#endif      /* LWM2M_BOOTSTRAP */
+            {
+                g_quit = 1;
+                continue;
+            }
+        }
+#ifdef LWM2M_BOOTSTRAP
+        DEBUG("update_bootstrap_info()\r\n");
+        update_bootstrap_info(&previousState, lwm2mH);
+#endif
+
+#ifdef MODULE_TINYDTLS
+handshake:
```

Comment:
The goto was there to skip the `lwm2m_step()` call while the DTLS handshake was still going on, as Wakaama would try to send data when there was no DTLS channel yet, causing it to assume the server wasn't answering, which would make it go into the `STATE_BOOTSTRAP_REQUIRED` state. I think this might've been fixed by switching from GNRC to the sock API and I just never tested it without this dirty workaround.

Reply:
Since it's a loop, changing the order would definitely be possible, even though looks kinda strange to wait for a reply before sending things (at least in the first iteration of the loop). But checking `!connP->noSec` doesn't actually do what you expect here, I think. That flag just indicates that DTLS should not be used for the current connection, so your proposed check would break any non-DTLS connections.

There would have to be a check on whether or not the handshake is complete, which would've been [this](https://github.com/RIOT-OS/RIOT/pull/8610#discussion_r187964140) part.

I don't think it matters anymore, though, as it works now without this workaround (on my end at least).

> Also, `tv` is still valid ?

I'm not sure what you mean by this. `lwm2m_step()` will set `tv` according to some internal logic to define a timeout for responses to packets it sent out.